### PR TITLE
Less respawn Time on killing rax

### PR DESCRIPTION
**Pull request by Niker323**
https://github.com/dota2unofficial/12v12/issues/59

### CODE REVIEW

**Comment by darklordabc** [comment]
Please remove prints from code, clean it up a bit. Also, can you add notifications for all teams? Something like "Radiant team has destroyed Dire's Ranged Barracks, they now have 1 second less respawn time" 
Also, can you have it so there is a little text popup above the destroyed building that is "-1" or "-2"

File: game/scripts/vscripts/addon_game_mode.lua [hunk 1]
```diff
@@ -172,7 +208,6 @@ function CMegaDotaGameMode:OnEntityKilled( event )
 			end
 		end
 		if player_id ~= -1 then
-			local name = killedUnit:GetUnitName()
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
```

File: game/scripts/vscripts/modifier_rax_bonus.lua [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+function modifier_rax_bonus:IsPurgable() return false end
+function modifier_rax_bonus:IsPurgeException() return false end
+function modifier_rax_bonus:RemoveOnDeath() return false end
+
+function modifier_rax_bonus:DeclareFunctions()
+	return { MODIFIER_PROPERTY_RESPAWNTIME }
+end
+
+function modifier_rax_bonus:GetModifierConstantRespawnTime()
+	if IsServer() then
```
Comment (ark120202) [inline game/scripts/vscripts/modifier_rax_bonus.lua]: You can wrap whole `modifier_rax_bonus:GetModifierConstantRespawnTime` definition instead. It also can include `modifier_rax_bonus:DeclareFunctions`.

File: game/scripts/vscripts/addon_game_mode.lua [hunk 3]
```diff
@@ -73,6 +73,9 @@ function CMegaDotaGameMode:InitGameMode()
 		end
 	end, nil)
 
+	_G.goodraxbonus = 0
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: Assuming existence of only 2 teams isn't extensible, instead it should be stored in something like `_G.raxBonuses = {}` where key is team and value is reduction

File: game/scripts/vscripts/modifier_rax_bonus.lua [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+		if self:GetParent():GetTeam() == DOTA_TEAM_GOODGUYS then
```
Comment (ark120202) [inline game/scripts/vscripts/modifier_rax_bonus.lua]: Would be
```lua
return -(raxBonuses[self:GetParent():GetTeam()] or 0)
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 5]
```diff
@@ -160,6 +163,39 @@ function CMegaDotaGameMode:OnEntityKilled( event )
     local killedUnit = EntIndexToHScript( event.entindex_killed )
     local killer = EntIndexToHScript( event.entindex_attacker )
 	local killedTeam = killedUnit:GetTeam()
+	local name = killedUnit:GetUnitName()
+
+	local raxbonuses = {
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: With `raxBonuses` as a global that name would be confusing
```suggestion
	local raxRespawnTimeWorth = {
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 6]
```diff
@@ -160,6 +163,39 @@ function CMegaDotaGameMode:OnEntityKilled( event )
...
+		npc_dota_badguys_range_rax_top = -1,
+		npc_dota_badguys_melee_rax_top = -2,
+		npc_dota_badguys_range_rax_mid = -1,
+		npc_dota_badguys_melee_rax_mid = -2,
+		npc_dota_badguys_range_rax_bot = -1,
+		npc_dota_badguys_melee_rax_bot = -2,
+	}
+	if raxbonuses[name] ~= nil then
+		if raxbonuses[name] > 0 then
+			_G.badraxbonus = _G.badraxbonus + raxbonuses[name]
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```lua
local opposingTeam = killedUnit:GetOpposingTeamNumber()
raxBonuses[opposingTeam] = raxBonuses[opposingTeam] + raxRespawnTimeWorth[name]
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 7]
```diff
@@ -160,6 +164,40 @@ function CMegaDotaGameMode:OnEntityKilled( event )
...
+		npc_dota_badguys_range_rax_top = -1,
+		npc_dota_badguys_melee_rax_top = -2,
+		npc_dota_badguys_range_rax_mid = -1,
+		npc_dota_badguys_melee_rax_mid = -2,
+		npc_dota_badguys_range_rax_bot = -1,
+		npc_dota_badguys_melee_rax_bot = -2,
+	}
+	if raxRespawnTimeWorth[name] ~= nil then
+		local opposingTeam = killedUnit:GetOpposingTeamNumber()
+		if raxRespawnTimeWorth[name] > 0 then
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: There's no reason to separate it by negativeness now

File: game/scripts/vscripts/addon_game_mode.lua [hunk 8]
```diff
@@ -160,6 +164,40 @@ function CMegaDotaGameMode:OnEntityKilled( event )
...
+			raxBonuses[opposingTeam] = raxBonuses[opposingTeam] + raxRespawnTimeWorth[name]
+		else
+			raxBonuses[opposingTeam] = raxBonuses[opposingTeam] - raxRespawnTimeWorth[name]
+		end
+		SendOverheadEventMessage( nil, OVERHEAD_ALERT_MANA_LOSS, killedUnit, math.abs(raxRespawnTimeWorth[name]), nil )
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: With change above
```suggestion
		SendOverheadEventMessage( nil, OVERHEAD_ALERT_MANA_LOSS, killedUnit, raxRespawnTimeWorth[name], nil )
```

**Comment by darklordabc** [comment]
Change the Russian text so that it is different for when all barracks is destroyed.
image

**Comment by darklordabc** [comment]
I still need to add Chinese translations for these texts.

**Comment by darklordabc** [comment]
I made a mistake in my latest upload, I didn't resolve the conflict properly in this file
https://github.com/dota2unofficial/12v12/pull/66/files#diff-692f02b94270242be6b0e6531e752c8a

I ended up deleting a lot of the changes in master branch, need to resolve conflict properly.

**Comment by darklordabc** [comment]
This PR adds files related to personal courier, I dont know why.

**Pull request merged**